From 5b6217f13b702212442f604921723ec11ea46b77 Mon Sep 17 00:00:00 2001
From: "yi.liu" <yi.liu@amlogic.com>
Date: Wed, 15 Nov 2023 10:44:38 +0800
Subject: [PATCH 3/5] amlbian: Support upgrading uboot by install deb [1/1]

PD#SWPL-146130

Problem:
Different startup methods require support for uboot upgrade.

Solution:
Add kernel startup parameters(boot_soure),
distinguishing the startup method of Debian.

Verify:
local

Change-Id: I95f5c4e73a8ca97f849c3bb42696b8b9a4c4526c
Signed-off-by: yi.liu <yi.liu@amlogic.com>
---
 board/amlogic/common/board.c                  | 59 +++++++++++++++++++
 board/amlogic/configs/s4_ap222_debian.h       |  5 +-
 board/amlogic/configs/sc2_ah212_debian.h      |  5 +-
 .../amlogic/configs/t7_an400_lpddr4x_debian.h |  4 +-
 board/amlogic/s4_ap222/s4_ap222.c             |  5 ++
 board/amlogic/sc2_ah212/sc2_ah212.c           |  5 +-
 .../t7_an400_lpddr4x/t7_an400_lpddr4x.c       |  4 ++
 include/amlogic/board.h                       |  1 +
 8 files changed, 81 insertions(+), 7 deletions(-)

diff --git a/board/amlogic/common/board.c b/board/amlogic/common/board.c
index dd5f87730a9..4cf9f209f5a 100644
--- a/board/amlogic/common/board.c
+++ b/board/amlogic/common/board.c
@@ -18,6 +18,7 @@
 #include <amlogic/aml_v2_burning.h>
 #include <amlogic/board.h>
 #include <asm/arch/secure_apb.h>
+#include <asm/arch/romboot.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 #define UNUSED(x) (void)(x)
@@ -301,3 +302,61 @@ int aml_board_late_init_tail(void *arg)
 }
 #endif//#ifdef CONFIG_BOARD_LATE_INIT
 
+static int board_get_bootid(void)
+{
+	const cpu_id_t cpuid = get_cpu_id();
+	const int familyId	 = cpuid.family_id;
+	int boot_id = 0;
+
+#ifdef SYSCTRL_SEC_STATUS_REG2
+	if (MESON_CPU_MAJOR_ID_SC2 <= familyId && MESON_CPU_MAJOR_ID_C2 != familyId &&
+		familyId != MESON_CPU_MAJOR_ID_T5W) {
+		boot_id = readl(SYSCTRL_SEC_STATUS_REG2);
+		boot_id = (boot_id >> 4) & 0xf;
+	}
+#endif// #ifdef SYSCTRL_SEC_STATUS_REG2
+
+#if defined(P_AO_SEC_GP_CFG0)
+	if ((familyId <= MESON_CPU_MAJOR_ID_T5D &&
+		familyId != MESON_CPU_MAJOR_ID_SC2) ||
+		familyId == MESON_CPU_MAJOR_ID_TXHD2 ||
+		familyId == MESON_CPU_MAJOR_ID_T5W) {
+		boot_id = readl(P_AO_SEC_GP_CFG0) & 0xf;
+	}
+#endif// #if defined(P_AO_SEC_GP_CFG0)
+
+	return boot_id;
+}
+
+void board_set_boot_source(void)
+{
+	const char *source;
+
+	switch (board_get_bootid()) {
+	case BOOT_ID_EMMC:
+		source = "emmc";
+		break;
+
+	case BOOT_ID_NAND:
+		source = "nand";
+		break;
+
+	case BOOT_ID_SPI:
+		source = "spi";
+		break;
+
+	case BOOT_ID_SDCARD:
+		source = "sd";
+		break;
+
+	case BOOT_ID_USB:
+		source = "usb";
+		break;
+
+	default:
+		source = "unknown";
+	}
+
+	env_set("boot_source", source);
+	printf("%s, boot_source: %s\n", __func__, source);
+}
diff --git a/board/amlogic/configs/s4_ap222_debian.h b/board/amlogic/configs/s4_ap222_debian.h
index 292ced108bf..6e6a8b430a1 100644
--- a/board/amlogic/configs/s4_ap222_debian.h
+++ b/board/amlogic/configs/s4_ap222_debian.h
@@ -146,7 +146,7 @@
 		"console=ttyS0,921600 console=tty0 no_console_suspend "\
 		"earlycon=aml-uart,0xfe07a000 ramoops.pstore_en=1 ramoops.record_size=0x8000 "\
 		"ramoops.console_size=0x4000 loop.max_part=4 scramble_reg=0xfe02e030 "\
-		"scsi_mod.scan=async xhci_hcd.quirks=0x800000 gamma=0 boot_source=emmc "\
+		"scsi_mod.scan=async xhci_hcd.quirks=0x800000 gamma=0 "\
 		"cma_first_wm_low=on "\
 		"\0"\
 	"upgrade_check="\
@@ -155,7 +155,8 @@
 	"storeargs="\
 		"get_bootloaderversion;" \
 		"run storeargs_base;"\
-		"setenv bootargs ${bootargs} ${emmc_quirks} kvm-arm.mode=none init_on_alloc=0;"\
+		"setenv bootargs ${bootargs} ${emmc_quirks} kvm-arm.mode=none init_on_alloc=0 "\
+			"boot_source=${boot_source};"\
 		"run cmdline_keys;"\
 		"\0"\
 	"switch_bootmode="\
diff --git a/board/amlogic/configs/sc2_ah212_debian.h b/board/amlogic/configs/sc2_ah212_debian.h
index 2512496f820..2e7e1f20ffd 100644
--- a/board/amlogic/configs/sc2_ah212_debian.h
+++ b/board/amlogic/configs/sc2_ah212_debian.h
@@ -145,7 +145,7 @@
 		"console=ttyS0,921600 console=tty0 no_console_suspend "\
 		"earlycon=aml_uart,0xfe07a000 ramoops.pstore_en=1 ramoops.record_size=0x8000 "\
 		"ramoops.console_size=0x4000 loop.max_part=4 scsi_mod.scan=async "\
-		"xhci_hcd.quirks=0x800000 loglevel=4 scramble_reg=0xfe02e030 boot_source=emmc "\
+		"xhci_hcd.quirks=0x800000 loglevel=4 scramble_reg=0xfe02e030 "\
 		"gamma=0 "\
 		"\0"\
 	"upgrade_check="\
@@ -154,7 +154,8 @@
 	"storeargs="\
 		"get_bootloaderversion;" \
 		"run storeargs_base;"\
-		"setenv bootargs ${bootargs} cma_first_wm_low=on kvm-arm.mode=none init_on_alloc=0;"\
+		"setenv bootargs ${bootargs} cma_first_wm_low=on kvm-arm.mode=none "\
+			"init_on_alloc=0 boot_source=${boot_source};"\
 		"run cmdline_keys;"\
 		"\0"\
 	"switch_bootmode="\
diff --git a/board/amlogic/configs/t7_an400_lpddr4x_debian.h b/board/amlogic/configs/t7_an400_lpddr4x_debian.h
index 7bced9258d4..7b1e361743e 100644
--- a/board/amlogic/configs/t7_an400_lpddr4x_debian.h
+++ b/board/amlogic/configs/t7_an400_lpddr4x_debian.h
@@ -200,7 +200,7 @@
 		"earlycon=aml_uart,0xfe078000 ramoops.pstore_en=1 ramoops.record_size=0x8000 "\
 		"amlogic_board=t7_an400_lpddr4x_debian "\
 		"ramoops.console_size=0x4000 loop.max_part=4 scsi_mod.scan=async "\
-		"xhci_hcd.quirks=0x800000 loglevel=4 scramble_reg=0xfe02e030 boot_source=emmc "\
+		"xhci_hcd.quirks=0x800000 loglevel=4 scramble_reg=0xfe02e030 "\
 		"gamma=0 "\
 		"\0"\
 	"upgrade_check="\
@@ -210,7 +210,7 @@
 		"get_bootloaderversion;" \
 		"run storeargs_base;"\
 		"setenv bootargs ${bootargs} kvm-arm.mode=none init_on_alloc=0 "\
-			"nn_adj_vol=${nn_adj_vol};"\
+			"nn_adj_vol=${nn_adj_vol} boot_source=${boot_source};"\
 		"run storeargs_hdmitx;"\
 		"run cmdline_keys;"\
 		"\0"\
diff --git a/board/amlogic/s4_ap222/s4_ap222.c b/board/amlogic/s4_ap222/s4_ap222.c
index 6e1ebed4732..783eb92a5ea 100644
--- a/board/amlogic/s4_ap222/s4_ap222.c
+++ b/board/amlogic/s4_ap222/s4_ap222.c
@@ -146,6 +146,11 @@ int board_late_init(void)
 	printf("board late init\n");
 	aml_board_late_init_front(NULL);
 
+#ifdef CONFIG_S4_AP222_DEBIAN
+	// Set boot source
+	board_set_boot_source();
+#endif
+
 #ifdef CONFIG_AB_UPDATE
 	extern int ab_mode(void);
 	ab_mode();
diff --git a/board/amlogic/sc2_ah212/sc2_ah212.c b/board/amlogic/sc2_ah212/sc2_ah212.c
index c845ed34a9c..ba9a07ad99d 100644
--- a/board/amlogic/sc2_ah212/sc2_ah212.c
+++ b/board/amlogic/sc2_ah212/sc2_ah212.c
@@ -34,7 +34,6 @@
 #include <amlogic/media/vout/aml_cvbs.h>
 #endif
 #include <amlogic/board.h>
-
 #include "avb2_kpub.c"
 
 DECLARE_GLOBAL_DATA_PTR;
@@ -165,6 +164,10 @@ int board_late_init(void)
 	printf("board late init\n");
 	aml_board_late_init_front(NULL);
 
+#ifdef CONFIG_SC2_AH212_DEBIAN
+	// Set boot source
+	board_set_boot_source();
+#endif
 	/* reset vout init state */
 	run_command("setenv vout_init disable", 0);
 #ifdef CONFIG_AML_VPU
diff --git a/board/amlogic/t7_an400_lpddr4x/t7_an400_lpddr4x.c b/board/amlogic/t7_an400_lpddr4x/t7_an400_lpddr4x.c
index 35ababac409..eab6ae0e746 100644
--- a/board/amlogic/t7_an400_lpddr4x/t7_an400_lpddr4x.c
+++ b/board/amlogic/t7_an400_lpddr4x/t7_an400_lpddr4x.c
@@ -247,7 +247,11 @@ int board_late_init(void)
 	aml_board_late_init_front(NULL);
 
 #ifdef CONFIG_T7_AN400_LPDDR4X_DEBIAN
+	/* Select fdtfile */
 	SetCurrentDtbFile();
+
+	/* Set boot source*/
+	board_set_boot_source();
 #endif
 
 	board_boot_freertos();
diff --git a/include/amlogic/board.h b/include/amlogic/board.h
index 0fa09929148..9709cb93c75 100644
--- a/include/amlogic/board.h
+++ b/include/amlogic/board.h
@@ -9,3 +9,4 @@ int aml_board_late_init_front(void *arg);
 
 int aml_board_late_init_tail(void *arg);
 
+void board_set_boot_source(void);
-- 
2.25.1

